components/org.wso2.carbon.identity.context.rewrite.valve/src/main/java/org/wso2/carbon/identity/context/rewrite/valve/TenantContextRewriteValve.java

```diff
@@ -69,15 +72,22 @@ public void invoke(Request request, Response response) throws IOException, Servl
         String contextToForward = null;
         boolean isContextRewrite = false;
         boolean isWebApp = false;
+        boolean isTenantDomainInThreadLocal = false;
 
         //Get the rewrite contexts and check whether request URI contains any of rewrite contains.
         for (RewriteContext context : contextsToRewrite) {
-            Pattern pattern = context.getPattern();
-            if (pattern.matcher(requestURI).find() || pattern.matcher(requestURI + "/").find()) {
+            Pattern patternTenant = context.getTenantContextPattern();
+            Pattern patternSuperTenant = context.getBaseContextPattern();
+            if (patternTenant.matcher(requestURI).find() || patternTenant.matcher(requestURI + "/").find()) {
                 isContextRewrite = true;
                 isWebApp = context.isWebApp();
                 contextToForward = context.getContext();
                 break;
+            } else if (!isTenantDomainInThreadLocal && patternSuperTenant.matcher(requestURI).find() ||
+                    patternSuperTenant.matcher(requestURI + "/").find()) {
+                String tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+                IdentityUtil.threadLocalProperties.get().put(TENANT_NAME_FROM_CONTEXT, tenantDomain);
+                isTenantDomainInThreadLocal = true;
```

Comment:
This would still set the thread domain to unexpected contexts.
For example suppose context is '/oauth2/' and the request comes as '/api/identity/oauth2/dcr'
The pattern will be matched and thread local will be set right

Rather you have to build the base pattern appending "^" so that at pattern match it checks whether request uri starts with the context configured